[RFC] Run all unit tests in separate processes

ZyX-I opened the pull request:
Should prevent all possible issues with state inherited from previous tests. Should also allow test to crash and still proceed with other tests.

Currently tests are also run with garbage collection disabled.

Note: tests are run in separate process. `before_each`, `after_each`, module code, etc is run in main process. So if you have conflicting imports or something like this this PR is not going to help.

Part of #5119.

After PR is merged it should also be possible to enable the same system which checks for core dumps which functional tests currently use, but I did not do that in this PR, primary because I do not need this much: except for rare occasions like with `tv_list_join()` unit tests normally either do not crash both locally and on CI or crash both locally and on CI.

ZyX-I committed:
unittests: Run tests in a separate process

ZyX-I committed:
unittests: Pause garbage collector while executing tests

Temporary (?) workaround for currently failing check_alloc_log tests.

ZyX-I committed:
unittests: Try using syscall library instead (ffi-based)

ZyX-I committed:
unittests: Do not use syscall library: does not work well with cimport

ZyX-I committed:
unittests: Use own bindings to libc syscall wrappers

ZyX-I committed:
unittests: Log syscalls if requested

ZyX-I committed:
unittests: Allow failing test to fail

ZyX-I committed:
unittests: Fix lint errors

ZyX-I committed:
unittests: Do not run failing test at all

ZyX-I committed:
unittests: Run all unit tests in their own processes

Used

    sed -r -i -e '/ helpers =/ s/$/\nlocal itp = helpers.gen_itp(it)/; s/^(\s*)it\(/\1itp(/' test/unit/**/*_spec.lua

to alter all tests. Locally they all run fine now.

Reasoning:

1. General: state from one test should not affect other tests.
2. Local: travis build is failing with something which may be an output of
   garbage collector. This should prevent state of the garbage collector from
   interferring as well.

justinmk commented:
macOS has some core dump: https://s3.amazonaws.com/archive.travis-ci.org/jobs/207806481/log.txt

justinmk commented:
After this change what happens if someone accidentally uses `it()` in a unit test? Should we make it an error?

ZyX-I commented:
> macOS has some core dump: https://s3.amazonaws.com/archive.travis-ci.org/jobs/207806481/log.txt

I saw, and there are way too much errors. Was going to try some Mac OS image on virtualbox, though one found on the [internet](https://techsviewer.com/how-to-install-mac-os-x-el-capitan-on-pc-on-virtualbox/) already fails with blank screen and virtualbox critical error.

> After this change what happens if someone accidentally uses it() in a unit test? Should we make it a testlint error?

Not sure how error may be done. Though nothing particularly bad is going to happen, just state may leak and affect other tests.

ZyX-I commented:
So far any attempt to run Mac OS in virtual machine results in that machine crashing, checked with different images and virtualbox or vmware.

ZyX-I committed:
unittests: Check core dumps in after_each, like in functests

ZyX-I committed:
unittests: Do not import libnvim or headers in main process

Slows down unit tests much, but gets rid of as much preserved state as possible.

ZyX-I committed:
unittests: Always close all pipes

ZyX-I committed:
unittests: Do not use which, add data to paths.lua.in instead

ZyX-I committed:
unittests: Update test/unit/message_spec.lua

ZyX-I commented:
@jamessan Any ideas why 9aa0132 is needed for test to pass?

ZyX-I committed:
unittests: Fix linter errors

ZyX-I committed:
tests: Allow lfs.attributes to be NULL when traversing FS

ZyX-I commented:
Got that working on travis OS X. And not working in other places: `lfs.attributes` appears to return `nil` randomly, like if somebody was cleaning up when it is running (should not be possible as long as main process uses `sc.wait(child_pid)`).

ZyX-I commented:
Tests passed, should be ready.

justinmk commented on test/unit/tempfile_spec.lua:
```diff
@@ -1,58 +1,64 @@
+  before_each(deferred_call(function()
+    lib.vim_deltempdir()
+  end))
+  separate_cleanup(function()
```
What is `separate_cleanup`? How is it different from `teardown` and `after_each`? A brief comment attached to `separate_cleanup` would help. Probably `deferred_call` also.

ZyX-I replied:
`separate_cleanup` is run inside a new (child) process, as well as `deferred_call`. I think that actually code needs some altering.

ZyX-I committed:
unittests: Use more adequate names for some functions

ZyX-I committed:
unittests: Move filtering cdefs to main process

ZyX-I committed:
unittests: Allow running `ffi.cdef` in the main process

justinmk commented:
Test lint:

```
make[4]: Entering directory `/home/travis/build/neovim/neovim/build'
Checking /home/travis/build/neovim/neovim/test/unit/helpers.lua [0m[31m[1m4 warnings[0m

    /home/travis/build/neovim/neovim/test/unit/helpers.lua:172:5: setting non-standard global variable [0m[1mcdefs[0m
    /home/travis/build/neovim/neovim/test/unit/helpers.lua:176:5: setting non-standard global variable [0m[1mcdefs[0m
    /home/travis/build/neovim/neovim/test/unit/helpers.lua:214:22: accessing undefined variable [0m[1mcdefs[0m
    /home/travis/build/neovim/neovim/test/unit/helpers.lua:215:7: accessing undefined variable [0m[1mcdefs[0m
```

ZyX-I committed:
tests: Add description of test directory structure and test env vars

ZyX-I committed:
unittests: Fix linter error
